subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c

```diff
@@ -674,97 +690,123 @@ static int firmware_block_received_cb(uint16_t obj_inst_id, uint16_t res_id, uin
+		LOG_INF("DFU is allocated already");
+		k_sem_give(&fw_buf_sem);
+		return -EAGAIN;
+	}
 
-	if (bytes_downloaded == 0) {
-		if (ongoing_obj_id != UNUSED_OBJ_ID) {
-			LOG_INF("DFU is allocated already");
-			return -EAGAIN;
-		}
-
-		ongoing_obj_id = obj_inst_id;
-		client_acknowledge();
-
+	if (push_data.bytes_downloaded == 0) {
 		image_type = dfu_target_img_type(data, data_len);
 		if (image_type == DFU_TARGET_IMAGE_TYPE_NONE) {
-			ret = -ENOMSG; /* Translates to unsupported image type */
-			goto cleanup;
+			k_sem_give(&fw_buf_sem);
+			return -ENOMSG; /* Translates to unsupported image type */
 		}
 		LOG_INF("Image type %d", image_type);
-		ret = fota_download_util_dfu_target_init(image_type);
+		push_data = (struct push_data){
+			.work = push_data.work,
+			.total_size = total_size,
+			.image_type = image_type,
+			.data = data,
+			.data_len = data_len,
+			.last_block = last_block,
+		};
+	} else {
+		if (push_data.return_code) {
+			/* If previous write was not succesfull, do not continue */
+			k_sem_give(&fw_buf_sem);
+			return push_data.return_code;
+		}
+		push_data.data = data;
+		push_data.data_len = data_len;
+		push_data.last_block = last_block;
+	}
+	ongoing_obj_id = obj_inst_id;
+
+	if (!last_block) {
+		/* Handle initialization, erasing and flash writing on background */
+		k_work_submit(&push_data.work);
+	} else {
+		/* Last block must be handled directly, so we get correct return code */
+		push_data_handler(&push_data.work);
+		return push_data.return_code;
+	}
+	return 0;
+}
+
+static void push_data_handler(struct k_work *work)
+{
+	uint8_t current_percent;
+	uint32_t current_bytes;
+	size_t dfu_offset = 0;
+	int ret = 0;
+
+	if (push_data.bytes_downloaded == 0) {
+		ret = fota_download_util_dfu_target_init(push_data.image_type);
 		if (ret) {
 			goto cleanup;
 		}
 
+		LOG_INF("%s firmware download started.",
+			push_data.image_type == DFU_TARGET_IMAGE_TYPE_MODEM_DELTA ||
+					push_data.image_type == DFU_TARGET_IMAGE_TYPE_FULL_MODEM
+				? "Modem"
+				: "Application");
+
 		/* Store Started DFU type */
-		target_image_type_store(obj_inst_id, image_type);
-		ret = dfu_target_init(image_type, 0, total_size, dfu_target_cb);
+		target_image_type_store(ongoing_obj_id, push_data.image_type);
+		ret = dfu_target_init(push_data.image_type, 0, push_data.total_size, dfu_target_cb);
 		if (ret < 0) {
 			LOG_ERR("Failed to init DFU target, err: %d", ret);
 			goto cleanup;
 		}
-
-		LOG_INF("%s firmware download started.",
-			image_type == DFU_TARGET_IMAGE_TYPE_MODEM_DELTA ||
-					image_type == DFU_TARGET_IMAGE_TYPE_FULL_MODEM ?
-				"Modem" :
-				"Application");
 	}
 
-	ret = dfu_target_offset_get(&offset);
+	ret = dfu_target_offset_get(&dfu_offset);
 	if (ret < 0) {
 		LOG_ERR("Failed to obtain current offset, err: %d", ret);
 		goto cleanup;
 	}
 
+	push_data.bytes_downloaded += push_data.data_len;
+
+	if (dfu_offset > push_data.bytes_downloaded) {
```

Comment:
So this means that we are retrying some data that was downloaded previously? Reading this I would expect that the download operation was started from the beginning as I do not understand how we would hit this otherwise.

It would be nice if the comment included why we ended up in the situation that we are. It kind of does, but could be more explicit.

Reply:
CoAP Block-Wise operation starts always from the beginning.
We have no way of telling server the offset where to skip to.

According to [rfc7959#section-2.9](https://datatracker.ietf.org/doc/html/rfc7959#section-2.9) we can only tell server to:
* "2.31 Continue", and it will send a next block.
* any other error code and the server stops sending.